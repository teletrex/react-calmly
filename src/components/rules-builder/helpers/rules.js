/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */



import update from 'immutability-helper';
import isArray from 'lodash/isArray';
import pick from 'lodash/pick';
// import uid from '@carbon/react';  //TODO:  Kinda odd way of doing this...

const uid = new Date().getTime();

export const findRule = (rule, ruleId) => {
  if (rule.id === ruleId)
    return {
      data: rule,
      trace: [],
    };

  if (!isArray(rule.data)) return {};

  const result = {
    data: null,
    trace: [],
  };

  rule.data.some((item, index) => {
    const { data, trace } = findRule(item, ruleId);
    if (!data) {
      return false;
    }

    result.data = data;
    result.trace = [...trace, { index, parent: rule }];

    return true;
  });

  return result;
};

/**
 * Generate an update object that will be used in update from 'immutability-helper'
 *
 * @param trace - foundRuleTrace from findRule
 * @param {Object} update
 * @param {Object} [update.updateItems={}]
 * @param {Object} [update.updateValues={}]
 * @param maxLevel - max level of nesting
 * @param level - current level
 */
const generateUpdateObject = (
  trace,
  { updateItems = {}, updateValues = {} },
  maxLevel,
  level = 0
) => {
  if (maxLevel === level)
    return { ...updateValues[level], data: { ...(updateItems[level] || {}) } };

  const newTrace = [...trace];
  const { index } = newTrace.pop();
  return {
    ...updateValues[level],
    data: {
      ...(updateItems[level] || {}),
      [index]: generateUpdateObject(newTrace, { updateItems, updateValues }, maxLevel, level + 1),
    },
  };
};

/**
 * Generate an update object that will be used in update from 'immutability-helper'
 * for updating 'level'
 */
const generateUpdateObjectForLevel = (rule, level) => {
  if (!isArray(rule.data)) return { level: { $set: level } };

  return {
    level: { $set: level },
    data: { ...rule.data.map(item => generateUpdateObjectForLevel(item, level + 1)) },
  };
};

/**
 * Update maxLevelInside of each Rule.
 * We need to do this when we delete some Rule.
 */
const updateMaxLevelInside = rule => {
  if (isArray(rule.data)) {
    let maxLevelInside = 0;
    const data = rule.data.map(item => {
      const ret = updateMaxLevelInside(item);
      maxLevelInside = Math.max(maxLevelInside, ret.maxLevelInside);
      return ret;
    });
    return { ...rule, maxLevelInside: maxLevelInside + 1, data };
  }

  return {
    ...rule,
    maxLevelInside: 0,
  };
};

export const updateRuleCriteria = (rule, criteriaKey, createRule) => {
  const newRule = update(rule.data, {
    criteriaKey: { $set: criteriaKey },
    data: { $set: createRule(criteriaKey, {}) },
  });

  const updateItems = {
    [rule.trace.length - 1]: {
      $splice: [[rule.trace[0].index, 1, newRule]],
    },
  };

  return generateUpdateObject(rule.trace, { updateItems }, rule.trace.length - 1);
};

export const insertNewRule = (rules, ruleId, ruleToInsert, direction) => {
  const rule = findRule(rules, ruleId);
  const { index } = rule.trace[0];

  const newRule = update(ruleToInsert, generateUpdateObjectForLevel(ruleToInsert, rule.data.level));

  const updateItems = {
    [rule.trace.length - 1]: {
      $splice: [[index + (direction === -1 ? 0 : 1), 0, newRule]],
    },
  };
  const updateValues = {
    ...new Array(rule.trace.length).fill(0).map((val, index) => ({
      operands: { $apply: x => x + newRule.operands },
      maxLevelInside: {
        $apply: x => Math.max(x, newRule.maxLevelInside + rule.trace.length - index),
      },
    })),
  };

  return generateUpdateObject(rule.trace, { updateItems, updateValues }, rule.trace.length);
};

export const addNewRule = (rules, groupId) => {
  const { data, trace } = findRule(rules, groupId);

  const newRule = {
    id: uid(),
    level: data.level + 1,
    maxLevelInside: 0,
    operands: 1,
    data: null,
  };

  const updateItems = {
    [trace.length]: { $push: [newRule] },
  };
  const updateValues = {
    ...new Array(trace.length + 1).fill({
      operands: { $apply: x => x + 1 },
    }),
  };

  return generateUpdateObject(trace, { updateItems, updateValues }, trace.length);
};

export const deleteRule = (rules, ruleId) => {
  const { data, trace } = findRule(rules, ruleId);
  if (!data) {
    return rules;
  }

  const { parent, index } = trace[0];

  if (parent.data.length === 1 && trace.length !== 1) {
    return deleteRule(rules, parent.id);
  }

  const updateItems = {
    [trace.length - 1]: {
      $splice: [[index, 1]],
    },
  };
  const updateValues = {
    ...new Array(trace.length).fill({
      operands: { $apply: x => x - data.operands },
    }),
  };

  const updateObject = generateUpdateObject(trace, { updateItems, updateValues }, trace.length - 1);
  return updateMaxLevelInside(update(rules, updateObject));
};

export const duplicateRule = ({ trace, data: rule }, createRule) => {
  const { index } = trace[0];
  const duplicateRule = {
    ...rule,
    data: createRule(
      rule.criteriaKey,
      pick(rule.data.ruleData, 'operator', 'value'),
      pick(rule.data.ruleTimeframe, 'operator', 'value')
    ),
    id: uid(),
  };

  const updateItems = {
    [trace.length - 1]: {
      $splice: [[index + 1, 0, duplicateRule]],
    },
  };
  const updateValues = {
    ...new Array(trace.length).fill({
      operands: { $apply: x => x + rule.operands },
    }),
  };

  return generateUpdateObject(trace, { updateItems, updateValues }, trace.length - 1);
};

export const changeRuleData = (rules, ruleId, ruleData) => {
  const rule = findRule(rules, ruleId);
  const { parent } = rule.trace[0];
  const updateItems = {};

  const newElements = parent.data.map(item =>
    item.id === ruleId ? update(item, { data: { ruleData: { $set: ruleData } } }) : item
  );

  updateItems[rule.trace.length - 1] = { $set: newElements };

  return generateUpdateObject(rule.trace, { updateItems }, rule.trace.length - 1);
};

export const changeRuleTimeframe = (rules, ruleId, timeframeData) => {
  const rule = findRule(rules, ruleId);
  const { parent } = rule.trace[0];
  const updateItems = {};

  const newElements = parent.data.map(item =>
    item.id === ruleId ? update(item, { data: { ruleTimeframe: { $set: timeframeData } } }) : item
  );

  updateItems[rule.trace.length - 1] = { $set: newElements };

  return generateUpdateObject(rule.trace, { updateItems }, rule.trace.length - 1);
};

export const changeConjunction = (rule, conjunction) => {
  const { index, parent } = rule.trace[0];

  const updateItems = {};
  let updateValues = {};
  if (conjunction.changeAll) {
    updateValues = {
      [rule.trace.length - 1]: { conjunction: { $set: conjunction.value } },
    };
  } else {
    const newGroup = {
      id: uid(),
      conjunction: conjunction.value,
      level: parent.data[index - 1].level,
      maxLevelInside:
        1 + Math.max(parent.data[index - 1].maxLevelInside, parent.data[index].maxLevelInside),
      operands: parent.data[index - 1].operands + parent.data[index].operands,
      data: [
        update(
          parent.data[index - 1],
          generateUpdateObjectForLevel(parent.data[index - 1], parent.data[index - 1].level + 1)
        ),
        update(
          parent.data[index],
          generateUpdateObjectForLevel(parent.data[index], parent.data[index].level + 1)
        ),
      ],
    };

    updateItems[rule.trace.length - 1] = { $splice: [[index - 1, 2, newGroup]] };
    updateValues = {
      ...new Array(rule.trace.length).fill(0).map((val, index) => ({
        maxLevelInside: {
          $apply: x => Math.max(x, newGroup.maxLevelInside + rule.trace.length - index),
        },
      })),
    };
  }

  return generateUpdateObject(rule.trace, { updateItems, updateValues }, rule.trace.length - 1);
};

export const groupRules = (rules, ruleId, oneConjunctionMode) => {
  const { data, trace } = findRule(rules, ruleId);
  const { index, parent } = trace[0];

  const newGroup = {
    id: uid(),
    conjunction: oneConjunctionMode || parent.conjunction,
    level: data.level,
    maxLevelInside: data.maxLevelInside + 1,
    operands: data.operands,
    data: [update(data, generateUpdateObjectForLevel(data, data.level + 1))],
  };

  const updateItems = {
    [trace.length - 1]: {
      $splice: [[index, 1, newGroup]],
    },
  };
  const updateValues = {
    ...new Array(trace.length).fill(0).map((val, index) => ({
      maxLevelInside: {
        $apply: x => Math.max(x, newGroup.maxLevelInside + trace.length - index),
      },
    })),
  };

  return generateUpdateObject(trace, { updateItems, updateValues }, trace.length - 1);
};

export const ungroupRules = (rules, groupId) => {
  const { data, trace } = findRule(rules, groupId);
  const { parent, index } = trace[0];

  if (!isArray(data.data)) {
    return ungroupRules(rules, parent.id);
  }

  const newElements = data.data.map(item =>
    update(item, generateUpdateObjectForLevel(item, data.level))
  );

  const updateItems = {
    [trace.length - 1]: {
      $splice: [[index, 1, ...newElements]],
    },
  };

  const updateObject = generateUpdateObject(trace, { updateItems }, trace.length - 1);

  return updateMaxLevelInside(update(rules, updateObject));
};

export const moveRule = (rules, ruleId, direction) => {
  const { data, trace } = findRule(rules, ruleId);
  const { parent, index } = trace[0];
  const limit = direction === 1 ? 0 : parent.data.length - 1;

  if (!data || (index === limit && trace.length === 1)) return rules;

  const updateItems = {};
  let groupIdToRemoveBracket = null;

  if (index !== limit) {
    const otherItemData = parent.data[index - direction].data;
    if (isArray(otherItemData)) {
      const newItem = update(data, generateUpdateObjectForLevel(data, otherItemData[0].level));
      updateItems[trace.length - 1] = {
        $splice: [[index, 1]],
        [index - direction]: {
          data: { $splice: [[direction === 1 ? Number.MAX_VALUE : 0, 0, newItem]] },
        },
      };
    } else {
      updateItems[trace.length - 1] = {
        $splice: [
          [index, 1],
          [index - direction, 0, data],
        ],
      };
    }
  } else {
    const { index: indexOfParent } = trace[1];
    const newItem = update(data, generateUpdateObjectForLevel(data, parent.level));

    updateItems[trace.length - 1] = { $splice: [[index, 1]] };
    updateItems[trace.length - 2] = {
      $splice: [[indexOfParent + (direction === 1 ? 0 : 1), 0, newItem]],
    };

    if (parent.data.length <= 1) groupIdToRemoveBracket = parent.id;
  }

  const updateObject = generateUpdateObject(trace, { updateItems }, trace.length - 1);
  const newRules = update(rules, updateObject);

  if (groupIdToRemoveBracket !== null) {
    return ungroupRules(newRules, groupIdToRemoveBracket);
  }

  return updateMaxLevelInside(newRules);
};
