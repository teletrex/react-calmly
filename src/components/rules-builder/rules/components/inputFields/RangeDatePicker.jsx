/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React, { useContext } from 'react';
import PropTypes from 'prop-types';

import { DatePicker, DatePickerInput } from '@carbon/react';
import RulesBuilderContext from '../../../RulesBuilderContext';
import { DATE_PATTERN, VALUES } from '../../constants';
import {
  getDayTimestamp,
  getMachineDateTime,
  getPrefixedClasses,
  getReadOnlyDate,
} from '../../helpers';

const RangeDatePicker = ({ id, onValueChange, value }) => {
  const { isReadOnly } = useContext(RulesBuilderContext);

  const handleOnChange = dates => {
    const [rangeStart, rangeEnd] = dates;

    if (rangeStart && rangeEnd) {
      onValueChange([getDayTimestamp(rangeStart), getDayTimestamp(rangeEnd)]);
    }
  };

  if (isReadOnly) {
    return (
      <>
        <time
          className={getPrefixedClasses([
            '--text-space-left',
            '--text-space-right',
            '--rule-value-readonly',
          ])}
          dateTime={getMachineDateTime(value[0])}
        >
          {getReadOnlyDate(value[0])}
        </time>
        -
        <time
          className={getPrefixedClasses(['--text-space-left', '--rule-value-readonly'])}
          dateTime={getMachineDateTime(value[1])}
        >
          {getReadOnlyDate(value[1])}
        </time>
      </>
    );
  }

  return (
    <DatePicker
      data-testid="input-field-range-date-picker"
      datePickerType="range"
      id={`${id}--input-field-date-picker`}
      onChange={handleOnChange}
      value={value}
    >
      <DatePickerInput id={`${id}--range-picker-value-1`} labelText="" pattern={DATE_PATTERN} />
      <DatePickerInput id={`${id}--range-picker-value-2`} labelText="" pattern={DATE_PATTERN} />
    </DatePicker>
  );
};

RangeDatePicker.propTypes = {
  id: PropTypes.string.isRequired,
  onValueChange: PropTypes.func.isRequired,
  value: PropTypes.arrayOf(PropTypes.number),
};

RangeDatePicker.defaultProps = {
  value: VALUES.UNPROVIDED,
};

export default RangeDatePicker;
