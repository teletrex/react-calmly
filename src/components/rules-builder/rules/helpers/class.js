/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import classNames from 'classnames';
import map from 'lodash/map';
import isArray from 'lodash/isArray';
import  settings  from '../../../../settings';

const { prefix } = settings;

export const getPrefixedClasses = (classes = []) =>
  classNames(map(isArray(classes) ? classes : [classes], className => `${prefix}${className}`));
