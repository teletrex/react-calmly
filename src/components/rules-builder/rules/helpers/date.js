/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




const getDate = timestamp => (timestamp ? new Date(timestamp) : new Date());

export const getDayTimestamp = datestring => {
  const day = getDate(datestring);
  return day.setHours(0, 0, 0, 0);
};

export const getTodayTimestamp = () => getDayTimestamp();

export const getReadOnlyDate = timestamp => {
  const date = getDate(timestamp);
  return `${date.getMonth() + 1}/${date.getDate()}/${date.getFullYear()}`;
};

// used for the datetime attribute in <time> tags
export const getMachineDateTime = timestamp => {
  const date = getDate(timestamp);
  return date.toISOString();
};
