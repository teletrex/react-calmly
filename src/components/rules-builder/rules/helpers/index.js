/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export * from './class';
export * from './date';
export * from './rule';
export * from './useFormValidation';
