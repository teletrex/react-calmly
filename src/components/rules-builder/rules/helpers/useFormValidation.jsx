/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import { useFormik } from 'formik';

export function useFormValidation({ initialValues, onSubmit, validate, validationSchema }) {
  return useFormik({
    initialValues,
    onSubmit,
    validate: validate || undefined,
    validateOnBlur: true,
    validateOnChange: false,
    validationSchema,
  });
}
