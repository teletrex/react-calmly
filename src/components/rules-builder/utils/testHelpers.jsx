/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';

import RulesBuilderContext from '../RulesBuilderContext';
import {
  ERROR_RELEGATION_REASON_TAG,
  relegateErrors,
} from '../../utils/test-helpers/relegate-errors';

export const datePickerErrorWrapper = fn => {
  relegateErrors(
    {
      reasonTag: ERROR_RELEGATION_REASON_TAG.FAILED_TO_RESOLVE,
      reason: "Couldn't trace the 'uncontrolled input error' in date component.",
      expectedErrorCount: 1,
    },
    () => {
      fn();
    }
  );
};

export const renderWithContext = (component, contextValue = {}) =>
  render(
    <RulesBuilderContext.Provider value={contextValue}>{component}</RulesBuilderContext.Provider>
  );
