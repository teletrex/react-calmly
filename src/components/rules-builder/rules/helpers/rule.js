/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import capitalize from 'lodash/capitalize';
import uniqueId from 'lodash/uniqueId';

import { composeOperators } from '../../utils/composeOperators';

export const getRuleOutput = (operator, value) => ({ operator, value });

export const getInitialOperator = (operatorValue, operators) =>
  operators.find(item => item.value === operatorValue);

export const getRuleConfigData = ruleConfig => {
  const { placeholder = {}, inputComponent } = ruleConfig;

  return {
    placeholder,
    ...(inputComponent && { inputComponent }),
  };
};

export const getRuleContainerConfigData = (ruleConfig, ruleData, operatorsList) => {
  const { disableCriteriaReselect, operators: ruleOperators = [] } = ruleConfig;
  const { value } = ruleData;
  const operators = ruleOperators.length
    ? composeOperators(ruleOperators, operatorsList)
    : operatorsList;
  let operator = getInitialOperator(ruleData.operator, operators);

  if (disableCriteriaReselect && !operator) {
    [operator] = operators;
  }

  return {
    operator,
    value,
    operators,
  };
};

export const getRuleId = ruleType => `${capitalize(ruleType)}Rule_${uniqueId()}`;
