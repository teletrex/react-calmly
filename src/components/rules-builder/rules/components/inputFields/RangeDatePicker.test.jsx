/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { getDayTimestamp } from '../../helpers';
import { datePickerErrorWrapper, renderWithContext } from '../../../utils/testHelpers';

import RangeDatePicker from './RangeDatePicker';

const initialDates = ['12/12/2020', '12/15/2020'];
const initialValues = initialDates.map(date => getDayTimestamp(date));

const props = {
  id: 'TestRangeDatepickerElement',
  onValueChange: jest.fn(),
};

describe('RangeDate Picker', () => {
  it('should render correctly', () => {
    const { container } = render(<RangeDatePicker {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with a provided value', () => {
    const { container } = render(<RangeDatePicker {...props} value={initialValues} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly for read-only', () => {
    const { container } = renderWithContext(<RangeDatePicker {...props} value={initialValues} />, {
      isReadOnly: true,
    });
    expect(container).toMatchSnapshot();
  });

  it('should set the rule value when a date is entered to date picker', () => {
    datePickerErrorWrapper(() => {
      const { getAllByPlaceholderText } = render(<RangeDatePicker {...props} />);
      const datePickerInputs = getAllByPlaceholderText('MM/DD/YYYY');

      fireEvent.change(datePickerInputs[0], { target: { value: initialDates[0] } });
      fireEvent.blur(datePickerInputs[0]);
      fireEvent.change(datePickerInputs[1], { target: { value: initialDates[1] } });
      fireEvent.blur(datePickerInputs[1]);

      expect(props.onValueChange).toHaveBeenCalledWith(initialValues);
    });
  });
});
