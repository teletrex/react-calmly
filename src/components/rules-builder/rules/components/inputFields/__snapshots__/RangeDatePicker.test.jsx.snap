// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeDate Picker should render correctly 1`] = `
<div>
  <div
    class="bx--form-item"
  >
    <div
      class="bx--date-picker bx--date-picker--range bx--date-picker--nolabel"
      data-testid="input-field-range-date-picker"
      id="TestRangeDatepickerElement--input-field-date-picker"
    >
      <span
        class="bx--date-picker-input__wrapper--relative"
      >
        <div
          class="bx--date-picker-container bx--date-picker--nolabel"
        >
          <div
            class="bx--date-picker-input__wrapper"
          >
            <input
              autocomplete="off"
              class="bx--date-picker__input flatpickr-input"
              id="TestRangeDatepickerElement--range-picker-value-1"
              pattern="d{1,2}/d{1,2}/d{4}"
              placeholder="MM/DD/YYYY"
              type="text"
              value=""
            />
            <svg
              aria-label="Open calendar"
              class="bx--date-picker__icon"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              role="img"
              style="will-change: transform;"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M13,2h-2V1h-1v1H6V1H5v1H3C2.4,2,2,2.4,2,3v10c0,0.6,0.4,1,1,1h10c0.6,0,1-0.4,1-1V3C14,2.4,13.6,2,13,2z M13,13H3V6h10V13z M13,5H3V3h2v1h1V3h4v1h1V3h2V5z"
              />
              <title>
                Open calendar
              </title>
            </svg>
          </div>
        </div>
        <svg
          aria-hidden="true"
          class="bx--date-picker-input__error bx--date-picker-input__error--hidden"
          focusable="false"
          height="16"
          preserveAspectRatio="xMidYMid meet"
          style="will-change: transform;"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 1C4.1 1 1 4.1 1 8s3.1 7 7 7 7-3.1 7-7-3.1-7-7-7zm2.7 10.5L4.5 5.3l.8-.8 6.2 6.2-.8.8z"
          />
          <path
            d="M10.7 11.5L4.5 5.3l.8-.8 6.2 6.2-.8.8z"
            data-icon-path="inner-path"
            fill="none"
            opacity="0"
          />
        </svg>
      </span>
      <span
        class="bx--date-picker-input__wrapper--relative"
      >
        <div
          class="bx--date-picker-container bx--date-picker--nolabel"
        >
          <div
            class="bx--date-picker-input__wrapper"
          >
            <input
              autocomplete="off"
              class="bx--date-picker__input"
              data-fp-omit=""
              id="TestRangeDatepickerElement--range-picker-value-2"
              pattern="d{1,2}/d{1,2}/d{4}"
              placeholder="MM/DD/YYYY"
              type="text"
              value=""
            />
            <svg
              aria-label="Open calendar"
              class="bx--date-picker__icon"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              role="img"
              style="will-change: transform;"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M13,2h-2V1h-1v1H6V1H5v1H3C2.4,2,2,2.4,2,3v10c0,0.6,0.4,1,1,1h10c0.6,0,1-0.4,1-1V3C14,2.4,13.6,2,13,2z M13,13H3V6h10V13z M13,5H3V3h2v1h1V3h4v1h1V3h2V5z"
              />
              <title>
                Open calendar
              </title>
            </svg>
          </div>
        </div>
        <svg
          aria-hidden="true"
          class="bx--date-picker-input__error bx--date-picker-input__error--hidden"
          focusable="false"
          height="16"
          preserveAspectRatio="xMidYMid meet"
          style="will-change: transform;"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 1C4.1 1 1 4.1 1 8s3.1 7 7 7 7-3.1 7-7-3.1-7-7-7zm2.7 10.5L4.5 5.3l.8-.8 6.2 6.2-.8.8z"
          />
          <path
            d="M10.7 11.5L4.5 5.3l.8-.8 6.2 6.2-.8.8z"
            data-icon-path="inner-path"
            fill="none"
            opacity="0"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;

exports[`RangeDate Picker should render correctly for read-only 1`] = `
<div>
  <time
    class="bx--text-space-left bx--text-space-right bx--rule-value-readonly"
    datetime="2020-12-12T06:00:00.000Z"
  >
    12/12/2020
  </time>
  -
  <time
    class="bx--text-space-left bx--rule-value-readonly"
    datetime="2020-12-15T06:00:00.000Z"
  >
    12/15/2020
  </time>
</div>
`;

exports[`RangeDate Picker should render correctly with a provided value 1`] = `
<div>
  <div
    class="bx--form-item"
  >
    <div
      class="bx--date-picker bx--date-picker--range bx--date-picker--nolabel"
      data-testid="input-field-range-date-picker"
      id="TestRangeDatepickerElement--input-field-date-picker"
    >
      <span
        class="bx--date-picker-input__wrapper--relative"
      >
        <div
          class="bx--date-picker-container bx--date-picker--nolabel"
        >
          <div
            class="bx--date-picker-input__wrapper"
          >
            <input
              autocomplete="off"
              class="bx--date-picker__input flatpickr-input"
              id="TestRangeDatepickerElement--range-picker-value-1"
              pattern="d{1,2}/d{1,2}/d{4}"
              placeholder="MM/DD/YYYY"
              type="text"
              value="12/12/2020"
            />
            <svg
              aria-label="Open calendar"
              class="bx--date-picker__icon"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              role="img"
              style="will-change: transform;"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M13,2h-2V1h-1v1H6V1H5v1H3C2.4,2,2,2.4,2,3v10c0,0.6,0.4,1,1,1h10c0.6,0,1-0.4,1-1V3C14,2.4,13.6,2,13,2z M13,13H3V6h10V13z M13,5H3V3h2v1h1V3h4v1h1V3h2V5z"
              />
              <title>
                Open calendar
              </title>
            </svg>
          </div>
        </div>
        <svg
          aria-hidden="true"
          class="bx--date-picker-input__error bx--date-picker-input__error--hidden"
          focusable="false"
          height="16"
          preserveAspectRatio="xMidYMid meet"
          style="will-change: transform;"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 1C4.1 1 1 4.1 1 8s3.1 7 7 7 7-3.1 7-7-3.1-7-7-7zm2.7 10.5L4.5 5.3l.8-.8 6.2 6.2-.8.8z"
          />
          <path
            d="M10.7 11.5L4.5 5.3l.8-.8 6.2 6.2-.8.8z"
            data-icon-path="inner-path"
            fill="none"
            opacity="0"
          />
        </svg>
      </span>
      <span
        class="bx--date-picker-input__wrapper--relative"
      >
        <div
          class="bx--date-picker-container bx--date-picker--nolabel"
        >
          <div
            class="bx--date-picker-input__wrapper"
          >
            <input
              autocomplete="off"
              class="bx--date-picker__input"
              data-fp-omit=""
              id="TestRangeDatepickerElement--range-picker-value-2"
              pattern="d{1,2}/d{1,2}/d{4}"
              placeholder="MM/DD/YYYY"
              type="text"
              value="12/15/2020"
            />
            <svg
              aria-label="Open calendar"
              class="bx--date-picker__icon"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              role="img"
              style="will-change: transform;"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M13,2h-2V1h-1v1H6V1H5v1H3C2.4,2,2,2.4,2,3v10c0,0.6,0.4,1,1,1h10c0.6,0,1-0.4,1-1V3C14,2.4,13.6,2,13,2z M13,13H3V6h10V13z M13,5H3V3h2v1h1V3h4v1h1V3h2V5z"
              />
              <title>
                Open calendar
              </title>
            </svg>
          </div>
        </div>
        <svg
          aria-hidden="true"
          class="bx--date-picker-input__error bx--date-picker-input__error--hidden"
          focusable="false"
          height="16"
          preserveAspectRatio="xMidYMid meet"
          style="will-change: transform;"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 1C4.1 1 1 4.1 1 8s3.1 7 7 7 7-3.1 7-7-3.1-7-7-7zm2.7 10.5L4.5 5.3l.8-.8 6.2 6.2-.8.8z"
          />
          <path
            d="M10.7 11.5L4.5 5.3l.8-.8 6.2 6.2-.8.8z"
            data-icon-path="inner-path"
            fill="none"
            opacity="0"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;
